libs/multihash/varint.hpp

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_MULTIHASH_VARINT_HPP
+#define IROHA_MULTIHASH_VARINT_HPP
+
+#include "interfaces/common_objects/byte_range.hpp"
+
+namespace iroha {
+  namespace multihash {
+
+    /**
+     * Try to read single varint int from buffer.
+     * https://github.com/multiformats/unsigned-varint#format
+     * @tparam NumberType some integer type
+     * @param[in|out] buffer which contains the varint int in the beginning. the
+     * variable will be overwritten with the rest of the data after the number
+     * just read.
+     * @param[out] number the read number on success.
+     * @return true if successfully read the number, false otherwise
+     */
+    template <typename NumberType>
+    inline bool readVarInt(shared_model::interface::types::ByteRange &buffer,
+                           NumberType &number) {
+      if (buffer.empty()) {
+        return false;
+      }
+      number = 0;
+      constexpr NumberType kSignificantBitsMask{0x7F};
+      constexpr std::byte kContinuationBitMask{0x80};
+      constexpr size_t kMaxVarIntLength = 8;
+      for (size_t i = 0; i < kMaxVarIntLength && i < buffer.size(); i++) {
+        if (i >= buffer.size()) {
+          return false;
+        }
+        number |= ((static_cast<NumberType>(buffer[i]) & kSignificantBitsMask)
+                   << (7 * i));
+        if ((buffer[i] & kContinuationBitMask) == std::byte{0}) {
+          const size_t read_bytes = i + 1;
+          buffer = shared_model::interface::types::ByteRange{
+              buffer.data() + read_bytes, buffer.size() - read_bytes};
+          return true;
+        }
+      }
+      return false;
+    }
```

Comment:
``` c++
template <typename NumberType, typename Container>
inline bool readVarInt(Container &buffer, NumberType &number) {
    if (buffer.empty()) {
        return false;
    }
    number = 0;
    constexpr std::byte kSignificantBitsMask{0x7F};
    constexpr std::byte kContinuationBitMask{0x80};
    constexpr size_t kPayloadSize = ((sizeof(number) << size_t(3)) + size_t(6)) / size_t(7);

    auto const * const beg = &buffer[0];
    auto const * const end = std::min(&buffer[kPayloadSize], &buffer[buffer.size()]);
    auto const * ptr = beg;
    size_t counter = 0;

    do {
        number |= (static_cast<NumberType>(*ptr & kSignificantBitsMask) << (size_t(7) * counter++));
    } while(std::byte(0) != (*ptr++ & kContinuationBitMask) && ptr != end);

    auto const read = ptr - beg;
    buffer = Container (beg + read, beg + buffer.size()); // TODO: remove this to smth else
    return (*(--ptr) & kContinuationBitMask) == std::byte(0);
}```
In 'TODO' mark we need to fix. Most of all, if we use vector as container, may be it will be better to decode in reverse direction, to avoid container data recreation and memoves in the case of erase range from the begginning?

Reply:
thank you! I used your cycle with some minor readability improvements. I like that it has less conditional statements than my former variant. Pls take a look at 5493111.

Reply:
thank you!